core/vdbe/execute.rs

```diff
@@ -2184,120 +2205,153 @@ pub fn op_seek(
-                        return Ok(InsnFunctionStepResult::Step);
+                } else {
+                    op
+                };
+                let rowid = if matches!(original_value, Value::Null) {
+                    match actual_op {
+                        SeekOp::GE { .. } | SeekOp::GT => {
+                            state.pc = target_pc.as_offset_int();
+                            return Ok(InsnFunctionStepResult::Step);
+                        }
+                        SeekOp::LE { .. } | SeekOp::LT => {
+                            // No integers are < NULL, so jump to target
+                            state.pc = target_pc.as_offset_int();
+                            return Ok(InsnFunctionStepResult::Step);
+                        }
                     }
-                    SeekOp::LE { .. } | SeekOp::LT => {
-                        // No integers are < NULL, so jump to target
-                        state.pc = target_pc.as_offset_int();
-                        return Ok(InsnFunctionStepResult::Step);
+                } else {
+                    int_key
+                };
+                state.op_seek_state = OpSeekState::Seek {
+                    rowid,
+                    op: actual_op,
+                };
+                continue;
+            }
+            OpSeekState::Seek { rowid, op } => {
+                let seek_result = {
+                    let mut cursor = state.get_cursor(*cursor_id);
+                    let cursor = cursor.as_btree_mut();
+                    return_if_io!(cursor.seek(SeekKey::TableRowId(*rowid), *op))
+                };
+                let found = match seek_result {
+                    SeekResult::Found => true,
+                    SeekResult::NotFound => false,
+                    SeekResult::TryAdvance => {
+                        state.op_seek_state = OpSeekState::Advance { op: *op };
+                        continue;
                     }
+                };
+                if !found {
+                    state.pc = target_pc.as_offset_int()
+                } else {
+                    state.pc += 1
                 }
-            } else {
-                int_key
-            };
-            let mut cursor = state.get_cursor(*cursor_id);
-            let cursor = cursor.as_btree_mut();
-            let found = return_if_io!(cursor.seek(SeekKey::TableRowId(rowid), actual_op));
-
-            if !found {
-                target_pc.as_offset_int()
-            } else {
-                state.pc + 1
+                return Ok(InsnFunctionStepResult::Step);
             }
-        };
-        state.pc = pc;
+            OpSeekState::Advance { op } => {
+                let found = {
+                    let mut cursor = state.get_cursor(*cursor_id);
+                    let cursor = cursor.as_btree_mut();
+                    match op {
+                        SeekOp::GT | SeekOp::GE { eq_only: false } => return_if_io!(cursor.next()),
```

Comment:
This logic works, but it's not immediately obvious, so to save poor programmer cycles, I would add a comment here that: 1. we are at the boundary of two leaf pages right now, and 2. for the purposes of satisfying `GT/GE`, finding _any_ row with `next()` will work, because of the divider cell in between the two pages that lead us originally to the current child page in the first place:

```
GT 666?

            667
          /     \
 661,665           <anything here is GT 666>
```

And the same for `LT/LE` and `prev()`:

```
LT 666?

                         664
                      /     \
<anything here is LT 666>    666,667,668
 ```